### PR TITLE
Improve error messages for syntax errors

### LINKED ISSUE #5178: Uninformative error message for extra semicolon in method chain
It's common in JavaScript to find method call chains like the one below, particularly when using query libraries like Lazy.js or jQuery:
```js
doSomething()
    .doA()
    .doB()
    .doC();
```

If one accidentally adds an extra semicolon in the middle of such a chain--a mistake I myself make way too often--like so:
```js
doSomething()
    .doA()
    .doB();
    .doC();
```

The resulting error message is pretty much as vague as they come:
```
SyntaxError: Syntax error
```

### PR DESCRIPTION
**Pull request by rhuanjl**
Picking up on discussion in #5178 and also issue #767

This PR:
1. Adds support for %s in syntax error messages (though only up to 2 uses of it per message)
1. Adds facility for getting a string representation of the token being parsed, and the previous token
1. Updates some commonly encountered error messages that currently only say "Syntax error"

I'd like to follow up and improve some other error messages but this seemed big enough for one PR.

I'm slightly nervous that I may have done something wrong in my implementation of %s support - it works in my tests but I don't 100% understand all the relevant logic.

fixes #5178

### CODE REVIEW

File: lib/Parser/perrors.h [hunk 1]
```diff
@@ -109,3 +109,10 @@ LSC_ERROR_MSG(1093, ERRLabelBeforeClassDeclaration, "Labels not allowed before c
 LSC_ERROR_MSG(1094, ERRLabelFollowedByEOF, "Unexpected end of script after a label.")
 LSC_ERROR_MSG(1095, ERRFunctionAfterLabelInStrict, "Function declarations not allowed after a label in strict mode.")
 LSC_ERROR_MSG(1096, ERRAwaitAsLabelInAsync, "Use of 'await' as label in async function is not allowed.")
+
+LSC_ERROR_MSG(1100, ERRExperimental, "Use of disabled experimental feature")
+LSC_ERROR_MSG(1200, ERRKeywordAfter, "Unexpected keyword '%s' after '%s'")
+LSC_ERROR_MSG(1201, ERRTokenAfter, "Unexpected token '%s' after '%s'")
+LSC_ERROR_MSG(1202, ERRIdentifierAfter, "Unexpected identifier '%s' after '%s'")
+LSC_ERROR_MSG(1203, ERRInvalidIdentifier, "Unexpected invalid identifier '%s' after '%s'")
+LSC_ERROR_MSG(1205, ERRValidIfFollowedBy, "%s is only valid if followed by %s")
```
Comment (dilijev) [inline lib/Parser/perrors.h]: Maybe quotes around these `%s` for consistency?

File: test/es6/HTMLComments.js [hunk 2]
```diff
@@ -22,7 +22,8 @@ WScript.Echo("Code before CRLF--> is reachable");
 --> WScript.Echo("Code after CRLF--> is unreachable");
 
 // CR
-WScript.Echo("Code before CR--> is reachable");--> WScript.Echo("Code after CR--> is unreachable");
+WScript.Echo("Code before CR--> is reachable");
+--> WScript.Echo("Code after CR--> is unreachable");
```
Comment (dilijev) [inline test/es6/HTMLComments.js]: These carriage return (CR, `\r`, or `0x0d`) characters were deliberately placed here without newline (`\n` or `0x0a`) and are part of the parser test. Can you revert these changes?

File: test/es6/HTMLComments.js [hunk 3]
```diff
@@ -39,10 +40,12 @@ WScript.Echo("Code before PS--> is reachable");
...
+
--> WScript.Echo("Code after CRLS--> is unreachable");
 
 // CRPS
-WScript.Echo("Code before CRPS--> is reachable");

--> WScript.Echo("Code after CRPS--> is unreachable");
+WScript.Echo("Code before CRPS--> is reachable");
+
--> WScript.Echo("Code after CRPS--> is unreachable");
```
Comment (dilijev) [inline test/es6/HTMLComments.js]: (same as above)

File: test/es6/HTMLComments.js [hunk 4]
```diff
@@ -39,10 +40,12 @@ WScript.Echo("Code before PS--> is reachable"); --> WScript.Echo("Code after P
  */
 
 // CRLS
-WScript.Echo("Code before CRLS--> is reachable"); --> WScript.Echo("Code after CRLS--> is unreachable");
+WScript.Echo("Code before CRLS--> is reachable");
+ --> WScript.Echo("Code after CRLS--> is unreachable");
```
Comment (dilijev) [inline test/es6/HTMLComments.js]: (same as above)

File: test/es6/HTMLComments.js [hunk 5]
```diff
@@ -71,4 +74,4 @@ assert.throws(function () { eval('/* */--> WScript.Echo("Code after /* */--> is
 var a = 1; a-->a; WScript.Echo("Code after post-decrement with a greater-than comparison (-->) is reachable");
 assert.areEqual(0, a, "Post decrement executes");
 
-assert.throws(function () { eval('/* */ --->'); }, SyntaxError, "HTMLCloseComment causes syntax error with an extra -", "Syntax error");
+assert.throws(function () { eval('/* */ --->'); }, SyntaxError, "HTMLCloseComment causes syntax error with an extra -", "Unexpected token '>' after '-'");
```
Comment (dilijev) [inline test/es6/HTMLComments.js]: These error messages are perhaps more descriptive of the parser state but probably just as confusing to the author. IMO this is fine though.

File: lib/Parser/perrors.h (hunk 1 above)
Comment (dilijev) [reply lib/Parser/perrors.h]: Nevermind, I see how this message is being used now. LGTM

File: test/es6/HTMLComments.js (hunk 2 above)
Comment (rhuanjl) [reply test/es6/HTMLComments.js]: I don't recall editing that line possibly an oddity of my vscode settings when I made the change further down - will revert with a hex editor.

File: test/es6/HTMLComments.js (hunk 5 above)
Comment (rhuanjl) [reply test/es6/HTMLComments.js]: Yeah I didn't like that one much but wasn't sure what to do with it; the parser has -- i.e. predecrement operator then a - i.e. unary minus then is expecting a statement but hits > I could special case it but would be a bit messy and wasn't sure it was worth it. At least with this option the author knows that it's the "-" and ">" that there's a problem with rather than looking earlier/later in the line.

Comment (dilijev) [reply test/es6/HTMLComments.js]: Probably not worth it. I think this is fine.

File: test/es6/HTMLComments.js (hunk 2 above)
Comment (dilijev) [reply test/es6/HTMLComments.js]: Yeah this is one of those things that editors will generally try to make sane. Getting the file into this weird line ending state is almost a matter of just editing the bytes.

**Comment by rhuanjl** [comment]
Two things I didn't highlight in my description above:
1. I intentionally left a gap in the error numbers in perrors.h before the ones I added with a thought that these more generic errors should stay at the bottom - hence leaving space for new specific ones before them. In hindsight not sure why I didn't put a comment in for this.
2. There are 4 calls to processError that I haven't updated to handle the parameters for %s they're in ScriptContext.cpp, BytecodeGenerator.cpp and SourceTextModuleRecord.cpp. I couldn't work out if parse errors could reach these points - if yes I need to update them or the error messages won't show correctly for those paths.

**Comment by dilijev** [comment]
> 2\. if yes I need to update them or the error messages won't show correctly for those paths.

IMO We should assume this can happen (otherwise it's dead code, right?) and get those messages updated.

**Comment by rhuanjl** [comment]
I have added the comment I mentioned above to perrors.h and updated to latest.

And I've re-examined all calls to ProcessException() and concluded that the others don't need updating to handle %s.
These are:
1. In SourceTextModuleRecord.cpp: Used specifically to handle OutOfMemory and StackOverFlow during parsing but not syntax errors which follow a different path.
2. In ByteCodeGenerator.cpp: Used for exceptions whilst generating byte code, not syntax errors.
3. 2 Calls in ScriptContext.cpp used in SerialiseScript and SerializeParserState both of these don't appear to be designed to handle syntax errors but rather out of memory and stackoverflow - though there's possibly a logical error here - when you hit a syntax error in SerializeParserState whilst testing with ch and -UseParserStateCache you get two error messages printed (one meaningful and one not) - either way I think this is a different issue.

Good to merge?

File: lib/Parser/cmperr.cpp [hunk 6]
```diff
@@ -4,8 +4,16 @@
 //-------------------------------------------------------------------------------------------------------
 #include "ParserPch.h"
 
-ParseExceptionObject::ParseExceptionObject(HRESULT hr) : m_hr(hr)
+ParseExceptionObject::ParseExceptionObject(HRESULT hr, LPCWSTR stringOneIn, LPCWSTR stringTwoIn)
 {
+    m_hr = hr;
```
Comment (dilijev) [inline lib/Parser/cmperr.cpp]: why move this assignment from the constructor-initialization syntax to the body of the constructor?

Comment (rhuanjl) [reply lib/Parser/cmperr.cpp]: That wasn't a new change. There was no real reason for it though, I thought it looked neater next to the two string properties, want me to revert it?

Comment (dilijev) [reply lib/Parser/cmperr.cpp]: I don't **think** it makes any difference. I guess we don't have a code guideline or real consistency around this point AFAICT. @Penguinwizzard do you know if this matters in terms of code gen?

You can leave it alone for now, so we can let the CI finish and merge this if it doesn't make a difference.

**Comment by dilijev** [comment]
I'm running internal verification tests and will merge when they complete.

File: lib/Parser/cmperr.cpp (hunk 6 above)
Comment (Penguinwizzard) [reply lib/Parser/cmperr.cpp]: It's mostly stylistic, and there's no real enforced style in ChakraCore (yet, anyway). I'd be decently surprised if we get different codegen due to this in optimized builds.

**Comment by dilijev** [comment]
Some internal test baselines affected by the change of error messages need updating. I'll take care of that and then merge the PR. Thanks @rhuanjl!
